File: coding/Leetcode/Biweekly_Leetcode/Biweekly_Contest_107/task_6898.py
from typing import List


class Solution:
    def join(self, x, y):
        if x[-1] == y[0]:
            return x + y[1:]
        else:
            return x + y

    def min_length(self, words):
        n = len(words)
        dp = [[0] * n for _ in range(n)]
        for i in range(n - 1, -1, -1):
            for j in range(i + 1, n):
                if i == j - 1:
                    dp[i][j] = len(self.join(words[i], words[j]))
                else:
                    dp[i][j] = float('inf')
                    for k in range(i, j):
                        dp[i][j] = min(dp[i][j], dp[i][k] + dp[k + 1][j])
                    dp[i][j] = min(dp[i][j], len(self.join(words[i], dp[i + 1][j])))
                    dp[i][j] = min(dp[i][j], len(self.join(dp[i][j - 1], words[j])))
        return dp[0][n - 1]

    def minimizeConcatenatedLength(self, words: List[str]) -> int:
        answer = self.min_length(words)
        return answer


def main():
    solve = Solution()

    words = ["aa", "ab", "bc"]
    answer = 4
    assert answer == solve.minimizeConcatenatedLength(words)

    words = ["ab", "b"]
    answer = 2
    assert answer == solve.minimizeConcatenatedLength(words)

    words = ["aaa", "c", "aba"]
    answer = 6
    assert answer == solve.minimizeConcatenatedLength(words)


if __name__ == "__main__":
    main()

File: coding/Leetcode/Biweekly_Leetcode/Biweekly_Contest_107/task_6898.cpp
//https://leetcode.com/contest/biweekly-contest-107/problems/decremental-string-concatenation/
// Created by ruslan515 on 24.06.23.
//

#include <bits/stdc++.h>

using namespace std;

class Solution {
public:

    int minimizeConcatenatedLength(vector<string> &words) {
        int n = words.size();
        string s = words[0];
        for (int i = 1; i < n; i++) {
            string s1 = s + words[i];
            string s2 = words[i] + s;
            if (s1.back() == s2.front()) {
                if (s1.size() < s2.size()) {
                    s = s1.substr(0, s1.size() - 1);
                } else {
                    s = s2.substr(1);
                }
            } else {
                s = min(s1, s2);
            }
        }
        return s.size();
    }
};

//class Solution {
//public:
//    int maximumNumberOfStringPairs(vector<string> &words) {
//        int answer = 0;
//        int n = words.size();
//        string str = words{0};
//        string curr, subStr;
//        for (int i = 1; i < n; ++i) {
//            curr = words{i};
//            if (str{str.size() - 1} == curr{0}) {
//                subStr = curr.substr(1);
//                str += subStr;
//            }
//        }
//
//        return answer;
//    }
//};

int main() {
    vector<string> words;
    int answer;
    Solution solve;

    words = {"aa", "ab", "bc"};
    answer = 4;
    assert(answer == solve.minimizeConcatenatedLength(words));

    words = {"ab", "b"};
    answer = 2;
    assert(answer == solve.minimizeConcatenatedLength(words));

    words = {"aaa", "c", "aba"};
    answer = 6;
    assert(answer == solve.minimizeConcatenatedLength(words));

    return 0;
}

